dipy/nn/histo_resdnn.py

```diff
@@ -0,0 +1,245 @@
+            unreliable prediction outside the brain.
+
+        Returns
+        -------
+        pred_sh_coef : np.ndarray (x, y, z, M)
+            Predicted fODF (as SH). The volume has matching shape to the input
+            data, but with ``(sh_order + 1) * (sh_order + 2) / 2`` as a last
+            dimension.
+
+        """
+        if mask is None:
+            logger.warning('Mask should be provided to accelerate '
+                           'computation, and because predictions are '
+                           'not reliable outside of the brain.')
+            mask = np.sum(data, axis=-1)
+        mask = mask.astype(bool)
+
+        # Extract B0's and obtain a mean B0
+        b0_indices = np.where(gtab.bvals < 1)[0]
+        if not len(b0_indices) > 0:
+            raise ValueError('b0 must be present for DWI normalization.')
+        logger.info('b0 indices found are: {}'.format(b0_indices))
+
+        mean_b0 = data[..., b0_indices]
+        mean_b0 = np.mean(mean_b0, axis=-1)
+
+        # Detect number of b-values and extract a single shell of DW-MRI Data
+        unique_shells = np.sort(np.unique(gtab.bvals))
+        logger.info('Number of b-values: {}'.format(unique_shells))
+
+        # Extract DWI only
+        dw_indices = np.where(gtab.bvals == unique_shells[1])[0]
+        dw_data = data[..., dw_indices]
+        dw_bvecs = gtab.bvecs[dw_indices, :]
+
+        # Normalize the DW-MRI Data with the mean b0 (voxel-wise)
+        norm_dw_data = np.zeros(dw_data.shape)
+        for n in range(len(dw_indices)):
+            norm_dw_data[..., n] = np.divide(dw_data[..., n], mean_b0,
+                                             where=np.abs(mean_b0) > 0.000001)
+
+        # Fit SH to the raw DWI signal
+        h_sphere = HemiSphere(xyz=dw_bvecs)
+        dw_sh_coef = sf_to_sh(norm_dw_data, h_sphere, smooth=0.0006,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
+
+        # Flatten and mask the data (N, SH_SIZE) to facilitate chunks
+        ori_shape = dw_sh_coef.shape
+        flat_dw_sh_coef = dw_sh_coef[mask > 0]
+        flat_pred_sh_coef = np.zeros(flat_dw_sh_coef.shape)
+
+        count = len(flat_dw_sh_coef) // 1000
+        for i in range(count+1):
+            if i % 100 == 0 or i == count:
+                logger.info('Chunk #{} out of {}'.format(i, count))
+            tmp_sh = self.predict(flat_dw_sh_coef[(i)*1000:(i+1)*1000])
+
+            # Removing negative values from the SF
+            sphere = get_sphere('repulsion724')
+            tmp_sf = sh_to_sf(sh=tmp_sh, sphere=sphere,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
+            tmp_sf[tmp_sf < 0] = 0
+            tmp_sh = sf_to_sh(tmp_sf, sphere, smooth=0.0006,
+                              basis_type=self.basis_type,
+                              sh_order=self.sh_order)
```

Comment:
this is strange to me. Why do you need to convert it and then come back?

Reply:
The model does predict negative fodf value (as SF), simply because this was not a constrained included in the original training/model.

But from SH you cannot remove negative value on SF, so you need to evaluate the SH, threshold, then fit new SH again.

Sub-optimal, yes. But necessary (I think)